### PR TITLE
Add check for long version name

### LINKED ISSUE #2554: Adding a long appversion name breaks out of the container
### What is the expected behaviour?
The appversion name should be restricted to 255 characters and validation needs to exist both on frontend and backend also when a long text is provided it shouldn't break out of the container

### What is the current behaviour?
No validation present for the character length and the text is breaking out of the container

### Screenshots or Screencast
image

### PR DESCRIPTION
**Pull request by tanmoyopenroot**
#### Checklist
- [x] Includes tests
- [ ] Update documentation

#### Changes proposed in this pull request:
Add char length check on version creation in frontend and backend

#### Screenshot

Screenshot from 2022-05-09 19-31-16

### CODE REVIEW

File: frontend/src/Editor/AppVersionsManager.jsx [hunk 1]
```diff
@@ -68,33 +68,44 @@ export const AppVersionsManager = function AppVersionsManager({
...
+      toast.error('The name of version should not be empty');
+      return;
+    }
 
-            const latestVersion = data.versions.at(0);
-            setAppDefinitionFromVersion(latestVersion);
-            setEditingAppVersion(latestVersion);
-            setVersionName('');
-          });
+    const passCharLimit = trimmedVersionName.length > 255;
```
Comment (Navaneeth-pk) [inline frontend/src/Editor/AppVersionsManager.jsx:79]: It will be more readable if we can rename the variable to `passwordCharacterLimit`.

Comment (tanmoyopenroot) [reply frontend/src/Editor/AppVersionsManager.jsx:79]: Umm, this is not related to `password` but more of the app version name char limit. So maybe `charLimitReached`?

Comment (Navaneeth-pk) [reply frontend/src/Editor/AppVersionsManager.jsx:79]: oh sorry, I read it as "password character limit". `isWithinCharacterLimit` or something similar works?

Comment (tanmoyopenroot) [reply frontend/src/Editor/AppVersionsManager.jsx:79]: Yeah, this works.

File: frontend/src/Editor/AppVersionsManager.jsx [hunk 2]
```diff
@@ -68,33 +68,44 @@ export const AppVersionsManager = function AppVersionsManager({
...
-        })
-        .catch((_error) => {
-          setIsCreatingVersion(false);
-          toast.error('Oops, something went wrong');
-        });
-    } else {
-      toast.error('The name of version should not be empty');
-      setIsCreatingVersion(false);
+    if (passCharLimit) {
+      toast.error('The name of version cannot be more than 255 characters');
```
Comment (gondar00) [inline frontend/src/Editor/AppVersionsManager.jsx:82]: nit: 
```suggestion
      toast.error('The version name cannot be more than 255 characters');
```

Comment (tanmoyopenroot) [reply frontend/src/Editor/AppVersionsManager.jsx:82]: Sure, will do

Comment (tanmoyopenroot) [reply frontend/src/Editor/AppVersionsManager.jsx]: Done

File: frontend/src/Editor/AppVersionsManager.jsx (hunk 1 above)
Comment (tanmoyopenroot) [reply frontend/src/Editor/AppVersionsManager.jsx]: Done

File: server/src/services/apps.service.ts [hunk 3]
```diff
@@ -290,6 +290,18 @@ export class AppsService {
   }
 
   async createVersion(user: User, app: App, versionName: string, versionFromId: string): Promise<AppVersion> {
+    const isEmpty = !versionName.length;
+
+    if (isEmpty) {
+      throw new BadRequestException('The version name should not be empty');
+    }
+
+    const isWithinCharacterLimit = versionName.length <= 255;
```
Comment (akshaysasidrn) [inline server/src/services/apps.service.ts:299]: You can rather use validation on _version-create.dto.ts_ file with a decorator like:
```
@MaxLength(255, {
    message: 'The version name cannot be more than 255 characters',
  })
```

**Comment by tanmoyopenroot** [comment]
> Hey @tanmoyopenroot, will be able to work on it? We could re-assign otherwise.

Will try to resolve all the changes by tomorrow.

File: frontend/src/Editor/AppVersionsManager.jsx [hunk 4]
```diff
@@ -68,33 +68,44 @@ export const AppVersionsManager = function AppVersionsManager({
...
-          toast.success('Version Created');
+    const trimmedVersionName = versionName.trim();
+    const isEmpty = !trimmedVersionName.length;
 
-          appVersionService.getAll(appId).then((data) => {
-            setAppVersions(data.versions);
+    if (isEmpty) {
+      toast.error('The version name should not be empty');
+      return;
+    }
```
[the remark this answers is not in the document]
Comment (tanmoyopenroot) [reply frontend/src/Editor/AppVersionsManager.jsx]: Done

File: frontend/src/Editor/AppVersionsManager.jsx [hunk 5]
```diff
@@ -68,33 +68,44 @@ export const AppVersionsManager = function AppVersionsManager({
...
-          setIsCreatingVersion(false);
-        })
-        .catch((_error) => {
-          setIsCreatingVersion(false);
-          toast.error('Oops, something went wrong');
-        });
-    } else {
-      toast.error('The name of version should not be empty');
-      setIsCreatingVersion(false);
+    if (!isWithinCharacterLimit) {
```
[the remark this answers is not in the document]
Comment (tanmoyopenroot) [reply frontend/src/Editor/AppVersionsManager.jsx]: Done

File: server/src/services/apps.service.ts [hunk 6]
```diff
@@ -290,6 +290,18 @@ export class AppsService {
   }
 
   async createVersion(user: User, app: App, versionName: string, versionFromId: string): Promise<AppVersion> {
+    const isEmpty = !versionName.length;
+
+    if (isEmpty) {
```
[the remark this answers is not in the document]
Comment (tanmoyopenroot) [reply server/src/services/apps.service.ts]: Done

File: server/src/services/apps.service.ts [hunk 7]
```diff
@@ -290,6 +290,18 @@ export class AppsService {
...
+      throw new BadRequestException('The version name should not be empty');
+    }
+
+    const isWithinCharacterLimit = versionName.length <= 255;
+
+    if (!isWithinCharacterLimit) {
```
[the remark this answers is not in the document]
Comment (tanmoyopenroot) [reply server/src/services/apps.service.ts]: Done

File: frontend/src/_styles/theme.scss [hunk 8]
```diff
@@ -71,6 +71,7 @@ button {
     margin: 0px auto;
```
[the remark this answers is not in the document]
Comment (tanmoyopenroot) [reply frontend/src/_styles/theme.scss]: Done

File: server/src/services/apps.service.ts [hunk 9]
```diff
@@ -290,6 +290,14 @@ export class AppsService {
   }
 
   async createVersion(user: User, app: App, versionName: string, versionFromId: string): Promise<AppVersion> {
+    if (!versionName) {
```
Comment (gsmithun4) [inline server/src/services/apps.service.ts:293]: Do we need the validations here? since we have it in the [dto](https://github.com/ToolJet/ToolJet/pull/3014/files#diff-e6f45f54d8d800f55022f14e4be153a11cb7088acf919080a4439012c1517ddbR8)

File: frontend/src/_styles/theme.scss [hunk 10]
```diff
@@ -5520,13 +5538,13 @@ word-break: break-all;
   border-right: 3px solid transparent;
   border-bottom: 8px solid transparent;
   box-shadow: 0 0 0 2px,
-  inset 6px 4px 0 -4px,
-  inset -6px 4px 0 -4px;
+    inset 6px 4px 0 -4px,
```
Comment (gsmithun4) [inline frontend/src/_styles/theme.scss:5541]: Please reverted the unrelated css changes

**Review by gsmithun4 — CHANGES_REQUESTED** [review]

File: server/src/services/apps.service.ts (hunk 9 above)
Comment (tanmoyopenroot) [reply server/src/services/apps.service.ts:293]: Currently `createVersion` is invoked from `app controller` [Line](https://github.com/ToolJet/ToolJet/blob/develop/server/src/controllers/apps.controller.ts#L261) and `VersionCreateDto` is added to this controller but what if some other `service/controller` uses this `createVersion` function without proper `check/DTO` in those might cause some issue

**Comment by withshubh** [comment]
Hey @tanmoyopenroot 👋 
Please let us know if you're still working on it?

**Comment by withshubh** [comment]
Closing this PR as there was no update for more than a month.
